app/models/user.rb

```diff
@@ -1,16 +1,45 @@
 class User < ActiveRecord::Base
-  has_many :posts
-  validates :password, presence: true, length: {minimum: 8}
-  validates :username, presence: true, length: {minimum: 3}
-  validates :email, uniqueness: true, presence: true
+has_many :posts
+has_many :comments
+require 'bcrypt'
 =begin
-	User attributes
-
-	username: 	this is the username that the user goes by it can be used to log in. 
-	 			It showed be shown next to users posts and comments. S
-	 			Should be unique
-	password: 	personal password used to log in 
-	email: 		user's email addres it can be used to log in
+User attributes
+username: 	this is the username that the user goes by it can be used to log in. 
+	 		It showed be shown next to users posts and comments. S
+	 		Should be unique
+password: 	personal password used to log in 
+email: 		user's email addres it can be used to log in
 =end
 
+
+	EMAIL_REGEX =  %r{[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\z}i
+
+	validates :username, presence: true, uniqueness: true, length: { in: 3..20 }
+	validates :email, presence: true, uniqueness: true, format: EMAIL_REGEX 
+	
+	validates :password, confirmation: true, length: {in: 6..20}
+	validates :password_confirmation, presence: true
+	
+	has_secure_password	  
+
+	def self.get_username_by_id(id)
+		user = User.find_by_id(id)
+		if user
+			return user.username
+		else
+			return id
+		end
+	end
+	def self.authenticate_with_username_or_email(username_or_email="",login_password="")
+		if EMAIL_REGEX.match(username_or_email)
+			user = User.find_by_email(username_or_email)
+		else
+			user = User.find_by_username(username_or_email)
+		end
+		if user && user.authenticate(login_password)
```

Comment:
Style nitpick.

```rb
if user && user.authenticate(login_password)
   user
end
```

If you want to be extra snazzy use try.

```rb
if user.try :authenticate, login_password
   user
end
```

I can't wait for the elvis operator to ship.

```rb
if user.?authenticate login_password
   user
end
```